tools/api-key-rotation/api_key_rotation_checker/main.py

```diff
@@ -0,0 +1,314 @@
+                params={"access_token": access_token}
+                ).json()
+        except exceptions.PermissionDenied:
+            continue
+        # If API keys exist, proceed
+        if "keys" in apikeys:
+            # Access our nested keys
+            # so we can iterate through the list
+            apikeys = apikeys["keys"]
+            # For each key in our dictionaries
+            # (API keys are dictionaries)
+            for apikey in apikeys:
+                # Google removed the "createdBy" field
+                # so only legacy keys have it
+                if "createdBy" in apikey:
+                    # Create our API key object
+                    # if it has "createdBy"
+                    key_object = ApiKey(
+                        apikey["keyId"],
+                        apikey["displayName"],
+                        apikey["createdBy"],
+                        apikey["createTime"],
+                        project_id)
+                else:
+                    # Create our API key object
+                    # if it does NOT have "createdBy"
+                    key_object = ApiKey(
+                        apikey["keyId"],
+                        apikey["displayName"],
+                        "None",
+                        apikey["createTime"],
+                        project_id)
+
+                # We need to convert
+                # our creation time for comparison
+                converted_creation_date = time_convert(key_object)
+
+                # Extract API Key ID for logging
+                key_id = key_object.key_id
+
+                # If our key is older than x days (default 90)
+                # based on our compare_dates function
+                # add api key to appropriate variable container
+                logging.info("Checking API key: %s creation date.. \n", key_id)
+                # Convert to JSON for logging
+                key_object_json = key_object.to_json()
+                if compare_dates(converted_creation_date, rotation_date):
+                    keys_needing_rotation.append(key_object_json)
+                else:
+                    keys_not_needing_rotation.append(key_object_json)
+
+    # Format our API keys
+    keys_needing_rotation = "\n".join(keys_needing_rotation)
+    keys_not_needing_rotation = "\n".join(keys_not_needing_rotation)
+
+    # Once analysis is complete for all keys,
+    # log results
+    if keys_needing_rotation:
+        logging.warning(
+            "Found API keys older than %s days. Please rotate: \n%s \n", rotation_period, keys_needing_rotation) # pylint: disable = line-too-long
+
+    if keys_not_needing_rotation:
+        logging.info(
+            "The following API key(s) are not older than %s days: \n%s", rotation_period, keys_not_needing_rotation) # pylint: disable = line-too-long
+
+
+def create_token():
+    """
+    There is no SDK for API Keys as of May 2020.
```

Comment:
Is this still valid? We are on March 2021 :)

Reply:
haha good call, yes there is a SDK but it is private / invite only alpha. This script can / should be updated when the SDK is available for public usage.

https://cloud.google.com/sdk/gcloud/reference/alpha/services/api-keys

I updated the docstring to be more accurate. Thanks